crates/sui-bridge-indexer/src/eth_bridge_indexer.rs

```diff
@@ -0,0 +1,373 @@
+        eth_rpc_url: String,
+        bridge_metrics: Arc<BridgeMetrics>,
+        indexer_metrics: BridgeIndexerMetrics,
+    ) -> Result<Self, anyhow::Error> {
+        let bridge_address = EthAddress::from_str(&eth_sui_bridge_contract_address)?;
+        Ok(Self {
+            bridge_address,
+            eth_rpc_url,
+            bridge_metrics,
+            indexer_metrics,
+        })
+    }
+}
+
+#[async_trait]
+impl Datasource<EthData, PgBridgePersistent, ProcessedTxnData> for EthFinalizedDatasource {
+    async fn start_data_retrieval(
+        &self,
+        task_name: String,
+        starting_checkpoint: u64,
+        target_checkpoint: u64,
+    ) -> Result<(JoinHandle<Result<(), Error>>, Receiver<(u64, Vec<EthData>)>), Error> {
+        let eth_client = Arc::new(
+            EthClient::<Http>::new(
+                &self.eth_rpc_url,
+                HashSet::from_iter(vec![self.bridge_address]),
+                self.bridge_metrics.clone(),
+            )
+            .await?,
+        );
+
+        let provider = Arc::new(
+            Provider::<Http>::try_from(&self.eth_rpc_url)?
+                .interval(std::time::Duration::from_millis(2000)),
+        );
+
+        info!("Starting from finalized block: {}", starting_checkpoint);
+
+        let finalized_contract_addresses =
+            HashMap::from_iter(vec![(self.bridge_address, starting_checkpoint)]);
+
+        let (task_handles, mut eth_events_rx, _) =
+            EthSyncer::new(eth_client, finalized_contract_addresses)
+                .run(self.bridge_metrics.clone())
+                .await
+                .map_err(|e| anyhow!(format!("{e:?}")))?;
+
+        let (data_sender, data_receiver) = metered_channel::channel(
+            1000,
+            &mysten_metrics::get_metrics()
+                .unwrap()
+                .channel_inflight
+                .with_label_values(&[&task_name]),
+        );
+        let indexer_metrics = self.indexer_metrics.clone();
+        let handle = spawn_monitored_task!(async {
+            'outer: while let Some((_, _, logs)) = eth_events_rx.recv().await {
+                // group logs by block, BTreeMap is used here to keep blocks in ascending order.
+                let blocks = logs.into_iter().fold(
+                    BTreeMap::new(),
+                    |mut result: BTreeMap<_, Vec<_>>, log| {
+                        let block_number = log.block_number;
+                        result.entry(block_number).or_default().push(log);
+                        result
+                    },
+                );
+                for (block_number, logs) in blocks {
+                    if block_number > target_checkpoint {
+                        break 'outer;
+                    }
+                    let mut data = vec![];
+                    let Ok(Ok(Some(block))) = retry_with_max_elapsed_time!(
+                        provider.get_block(block_number),
+                        Duration::from_secs(30)
```

Comment:
use a larger value, this means it will retry for 30s, which is a bit short when it's being rate limited